### ISSUE TITLE
Support for azurerm_monitor_data_collection_rule.dataCollectionEndpointId

### ISSUE THREAD

**Issue by BlackCoffeeAR** [hashicorp/terraform-provider-azurerm#20542]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

It would be nice to be able to associate existing Data Collection Endpoint to the Data Collection Rule at rule creation.
[The REST API Data Collection Rules - Create](https://learn.microsoft.com/en-us/rest/api/monitor/data-collection-rules/create?tabs=HTTP)

> properties.dataCollectionEndpointId		
> string
> The resource ID of the data collection endpoint that this rule can be used with.

### New or Affected Resource(s)/Data Source(s)

azurerm_monitor_data_collection_rule

### Potential Terraform Configuration

```hcl
resource "azurerm_monitor_data_collection_rule" "example" {
  name                = "example"
  resource_group_name = "example"
...

  data_flow {
    streams      = [
        "example"
    ]
    destinations = [
        "example"
    ]
  }

  destinations {
    log_analytics {
      workspace_resource_id = "example"
      name                  = "example"
    }
  }
  data_collection_endpoint_id = example_id
}
```

### References

https://learn.microsoft.com/en-us/rest/api/monitor/data-collection-rules/create?tabs=HTTP

**Comment by teowa** [comment]
Hi @BlackCoffeeAR , thanks for submitting this feature requset.

By checking the doc, the provided feature is only available in API version **2021-09-01-preview** which is a preview API version. We always use the stable api-version, because terraform doesn’t want to have any breaking changes introduced in preview api-version. And once there is a stable API version supports this feature, we will add this.

Thanks.

**Comment by BlackCoffeeAR** [comment]
Hi @teowa, thank you for reply, I understood.

**Comment by dawsonar802** [comment]
@BlackCoffeeAR, I know you closed this, but did you find a work around for this at all?  I have tried with the `AzAPI` provider as well but still am not able to set the Data Collection Endpoint ID.  When deploying via the portal this seems to work fine and I can associate a resource with the endpoint.  When you go to export the template and then re-deploy it though, there is no endpoint associated.  I am going to open a MS Support case as well, just was curious if you happened to get this to work?

**Comment by BlackCoffeeAR** [comment]
> @BlackCoffeeAR, I know you closed this, but did you find a work around for this at all? 
Hi @dawsonar802,  yes I did it with `AzApi` and it is working.
Here is my ressource:

```hcl
resource "azapi_resource" "example" {
  type      = "Microsoft.Insights/dataCollectionRules@2021-09-01-preview"
  name      = "example"
  parent_id = "example_id"
  location  = "example_location"
  tags      = var.tags

  body = jsonencode({
    properties = {
      // set here the id of the DCE resource
      dataCollectionEndpointId = azapi_resource.example_dce.id
      description = "description"
      dataFlows = [
        {
          streams = [
            "Custom-table_CL"
          ]
          destinations = [
            "table"
          ]
          transformKql = "source | extend TimeGenerated = now()"
          outputStream = "Custom-table_CL"
        }
      ]
      destinations = {
        logAnalytics = [
          {
            workspaceResourceId = azurerm_log_analytics_workspace.example_workspace.id
            name = "table"
          }
        ]
      }
    }
  })

  depends_on = [azapi_resource.example_table]
}
```
Do you get an error during deploying with terraform? Which one?

The export template contains also the block `parameters` with DCE ID:

``` json
    "parameters": {
        ...
        "dataCollectionEndpoints_<name>_dce_externalid": {
            "defaultValue": "/subscriptions/<id>/resourceGroups/<rg>/providers/Microsoft.Insights/dataCollectionEndpoints/<name>-dce",
            "type": "String"
        },
        ...
    },
```

**Comment by dawsonar802** [comment]
Thanks for the reply.  I may have a bit of a different config, as I am setup up the rule and endpoint via one deployment and then was running another deployment to do the rule and endpoint association.  

I do not get any errors, but when looking at the rule in the portal after creation, the endpoint is is not set.  I do see the parameter and value in the template though.  I am replying via mobile now, but can add some screenshots or configs later if you would like to see them.  

I was able to run the normal azurerm for the rule creation and then did an azapi update to set the endpoint.  This again seems to update the correct spot in the template when I view it, but after doing an association the resource does not have the endpoint set.  Seems very odd to me.

**Comment by BlackCoffeeAR** [comment]
> when looking at the rule in the portal after creation, the endpoint is is not set.
 Are you looking at the JSON view? Did you changed the API version to "2021-09-01-preview" or "2022-06-01"?

**Comment by dawsonar802** [comment]
> > when looking at the rule in the portal after creation, the endpoint is is not set.
> 
> Are you looking at the JSON view? Did you change the API version to "2021-09-01-preview" or "2022-06-01" in the upper right corner?

Yeah, through there and from Export Template, it looks fine.  The properties show the `dataCollectionEndpointId` correct.  The issue is when I look at the following view, it is not associating the endpoint.  
 
![Screenshot 2023-03-03 183409](https://user-images.githubusercontent.com/62777308/222854746-ca5577d5-de03-48e8-a38c-693ce752dabd.png)

I was expecting the endpoint to be set here in the DCR Resources since I am doing an association as well to a VM.

**Comment by BlackCoffeeAR** [comment]
In this view (Resources) I have the same: "No endpoind required" (my DCE is present in the dropdown list). But the log ingestion is working fine for me. So either it is not important what is set dor DCE in tab "Resources" or maybe it is only not shown correctly.

**Comment by dawsonar802** [comment]
> In this view (Resources) I have the same: "No endpoind required" (my DCE is present in the dropdown list). But the log ingestion is working fine for me. So either it is not important what is set dor DCE in tab "Resources" or maybe it is only not shown correctly.

Ah, interesting.  I have not actually checked logs yet, been stuck trying to see why this was not showing, but it may be working as well like your.  I have a MS case open as well, so I will update here when/if I get a reply from them.  For now I will see if the logs are coming in.  It seems like if the resource is associated both with the rule and the endpoint it would work fine, which is my case, just very odd it doesn't show in that drop down.  If I create a new rule in the portal and add the endpoint it shows there just fine.  Very strange.

**Comment by dawsonar802** [comment]
Quick update in case anyone else comes across this or is following.  I was able to review the logs and it appears AMA and the collection rules/endpoint are working as they should.  The issue of the endpoint not showing in the portal seems like it may be due to the API version.  As you noted earlier @BlackCoffeeAR when looking at the json for the rule, you have to select the preview version of the API to see the property there, so it seems like the portal is just using the older API version and not able to display the property.  The odd thing is the dropdown for the endpoint shows them and they can be selected.

**Comment by dawsonar802** [comment]
Just an update for anyone that is following or comes across this in the future.  I am still working with MS support on why the DCE is not showing in the portal despite it showing in the JSON and in the Export Template.  The thought is this is likely due to the API version. 

Something else, a little off this topic, but still useful and may be good to note in the Terraform Docs (I will try and put in a request to update this), but when using the `Microsoft-WindowsEvent` stream, you must have Azure Sentinel enabled on the workspace when using DCE.  If this is not enabled, you will get an error about the table not being available.  I can provide an MS Support case as reference if needed.  

With regards to the API version, @teowa , would it be possible to look into adding this functionality mention in this post based on there being a new version of the API?  The API docs are only showing the preview version but in the portal, looking at json view for a rule, there is a 2022-06-01 version.  I asked the MS Support person for the docs to be updated on this and it sounds like the product group is in progress, but has no ETA.
